src/core/AutomationClip.cpp

```diff
@@ -804,6 +872,11 @@ void AutomationClip::loadSettings( const QDomElement & _this )
 {
 	QMutexLocker m(&m_clipMutex);
 
+	// Legacy compatibility: Previously tangents were not stored in
+	// the project file. So if any node doesn't have tangent information
+	// we will generate the tangents
+	bool shouldGenerateTangents = false;
```

Comment:
Why not just write ``bool shouldGenerateTangents = !element.hasAttribute("inTan") && !element.hasAttribute("outTan");`` and use ``!shouldGenerateTangents`` for line 906? Then you could just remove the else that comes after it.

Reply:
I think in this case it would behave slightly different. The idea there is that if there's a single node on the project XML that doesn't have tangent information, we should run `generateTangents()`. That would work even if some nodes have the info and some don't (that's an edge case that we won't run into unless someone manually modify a project file I think, but we end up covering for that possibility with not much cost in the code). If I just use `bool shouldGenerateTangents = !element.hasAttribute("inTan") && !element.hasAttribute("outTan");` then `shouldGenerateTangents` will only depend on the last node of the automation clip, ignoring all the others that were cycled (so if there are 5 nodes with no tangent information and the last has, it won't generate tangents for the first 5). What do you think? For now I didn't change that, but if it's still preferred I can change it quickly in a new commit.

Reply:
> What do you think? For now I didn't change that, but if it's still preferred I can change it quickly in a new commit.

Sorry if my suggestion changed the behavior somehow. I was under the assumption that ``shouldGenerateTangents`` would be set true if both ``inTan`` or ``outTan`` was not found as an attribute in ``element``. My main goal here was just to reduce verbosity. I will leave to you to do what is best for the PR and the expected behavior.

Reply:
It's been a while since I wrote the code, I had to read it to see if there was any difference so I could be wrong. But the `shouldGenerateTangents` is inside a `for` loop that goes through each node. Setting it as `bool shouldGenerateTangents = !element.hasAttribute("inTan") && !element.hasAttribute("outTan");` will set it as true/false based on the last iteration of the loop. In most normal cases that won't matter because if one node has tangent information, all others would normally have. But the behavior is slightly different as in how it's written now: if any of the nodes don't have tangent information, the tangents will be generated, even if some do have the information because inside the loop `shouldGenerateTangents` will have been set as true once. The only situation I could think of that happening is someone editing a project file XML (say, copying some automation from an older project to a new one), which is a very edge case scenario, so if it's preferred that I change it I'm fine with it.

One alternative I thought of right now would be to have: `shouldGenerateTangents = shouldGenerateTangents || (!element.hasAttribute("inTan") && !element.hasAttribute("outTan"));`. Do you think this one looks confusing for someone reading the code?

Reply:
Sorry, I see the for-loop now (should've probably reviewed this in my editor). You obviously can't access ``element`` out of it, so my suggestion doesn't have much merit. I wish we could use ``std::all_of`` or ``std::any_of``, but the Qt Xml API doesn't really have iterators as far as I am aware.

I think it's fine then. I don't really see a better solution. Another thing is though, is that if there were 5 nodes, and one of them was missing the ``inTan`` attribute but had ``outTan``, we would still call ``generateTangents()`` even if that node did have tangent information. I'm not sure if this is possible in practice, but in theory, it might be. I think you would want to check for ``outTan`` as well on line 901.

Reply:
Right now, when you edit either the `inTan` or the `outTan` both tangents become "locked" and are both saved in the XML (they're also both saved even if they are "unlocked"). So in practice there shouldn't be a node with one attribute but not the other. Unless someone decides to delete one of them but not the other on the XML, in which case I'm not sure but it could cause a bug. I could set a default value on the `element.attribute("inTan")` call, though I think by default it's an empty string, so `LocaleHelper::toFloat` would convert it to 0? I could use `element.attribute("inTan", "0");` instead

Reply:
> I could set a default value on the element.attribute("inTan") call, though I think by default it's an empty string, so LocaleHelper::toFloat would convert it to 0? I could use element.attribute("inTan", "0"); instead

I guess it's fine then. I also realized that ``element.attribute`` would return an empty string if the attribute is not found, and ``QLocale::setDouble`` would return ``"0"`` on failure.